python-package/lightgbm/basic.py

```diff
@@ -761,22 +750,40 @@ def __init__(self, data, label=None, reference=None,
         self.silent = silent
         self.feature_name = feature_name
         self.categorical_feature = categorical_feature
-        self.params = copy.deepcopy(params)
+        self.params = self._filter_params(params)
         self.free_raw_data = free_raw_data
         self.used_indices = None
         self.need_slice = True
         self._predictor = None
         self.pandas_categorical = None
         self.params_back_up = None
-        self.feature_penalty = None
-        self.monotone_constraints = None
 
     def __del__(self):
         try:
             self._free_handle()
         except AttributeError:
             pass
 
+    def _filter_params(self, params):
+        if params is None:
+            return params
+        keys = ['has_header', 'header', 'max_bin', 'max_bin_by_feature', 'bin_construct_sample_cnt', 'subsample_for_bin', 'min_data_in_bin', 'use_missing',\
```

Comment:
Is this [IO params](https://lightgbm.readthedocs.io/en/latest/Parameters.html#io-parameters) + `forcedbins_filename` + `min_data_in_leaf`? Then a lot of params are missed, e.g. `histogram_pool_size`, `two_round`, etc.

Reply:
I think histogram_pool_size is not related to the dataset loading

Reply:
@guolinke 
OK, I see.
What about `forcedbins_filename`? It seems to me that it can be moved under **IO** section.

Also, I'm worrying about `seed` (`data_seed`) and `nthreads` parameters. Will they work correctly after this PR?

Reply:
... as well as about `initscore_filename` and `valid_data_initscores` params.

Reply:
> I think histogram_pool_size is not related to the dataset loading

Can we have a subsection under `IO` params named `Dataset` (or somehow better) params? Yes, we have no `Booster`/`Dataset` abstraction at CLI level, but with this PR distinguishing becomes very important. Without such info, users have to understand where are Dataset params only via fatal errors they get.


Or maybe we should add a **Note** to each Dataset parameter about that it cannot be changed after dataset construction.

WDYT?

Reply:
@guolinke Can you please take a look at this thread? I think that is all things I'm worrying about within this PR.

Reply:
sorry, I didn't see this thread ...
I agree with the subsection for dataset object, it will be much clearer.

Reply:
However, how about changing it in another PR? as it is related to documentation.

Reply:
@guolinke Thank you for your answers!

> However, how about changing it in another PR? as it is related to documentation.

Agree! I will help with docs in a separate PR.

Can you please answer to one remained my question?

> Also, I'm worrying about `seed` (`data_seed`) and `nthreads` parameters. Will they work correctly after this PR?

> ... as well as about `initscore_filename` and `valid_data_initscores` params.

Reply:
I think `nthreads` is safe for it doesn't affect the result of dataset. 
For `seed`, I think only consider `data_seed` is enough.
For `init_scores`, I think there are not loaded by python/R package, as we set it to `""` in the code:
https://github.com/microsoft/LightGBM/blob/509c2e50c25eded99fc0997afe25ebee1b33285d/src/c_api.cpp#L481

Reply:
> I think nthreads is safe for it doesn't affect the result of dataset.

Hmm, but it is used in many places during creating of dataset, e.g.
https://github.com/microsoft/LightGBM/blob/509c2e50c25eded99fc0997afe25ebee1b33285d/src/io/sparse_bin.hpp#L82
https://github.com/microsoft/LightGBM/blob/dbb804f2350f280f3c927f792f7e55ac9c021707/src/io/dataset.cpp#L662
I suppose I'm missing something...

> For seed, I think only consider data_seed is enough.

`seed` will cause changing of `data_seed` (except the case when `rand.NextShort()` will choose the same value as it was) and in consequence trigger the error, right?

> For init_scores, I think there are not loaded by python/R package, as we set it to "" in the code:

Oh, then our docs are incorrect:
https://github.com/microsoft/LightGBM/blob/509c2e50c25eded99fc0997afe25ebee1b33285d/include/LightGBM/config.h#L534-L535
https://github.com/microsoft/LightGBM/blob/509c2e50c25eded99fc0997afe25ebee1b33285d/include/LightGBM/config.h#L542-L543

Reply:
for `num_threads`, there are some temporary buffer, to speed-up the data loading by multi-threading. with different num_threads, the constructed dataset should be identical. As for `hist_buf_`, it is used when tree learning, and it is a temporary buffer too.

For seed, yes, changing `seed` will also changing `data_seed`, and the error will be raised too.

Reply:
Thank you very much for the detailed explanation!

Is it possible to update code for `init_scores` to match described behavior in docs?

Reply:
unfortunately, It is not trivial. As we don't know the filename passed in c_api is training or k-th valid data. It is hard to retrieve the corresponding init score filename. The easier way in python/R is read the init_score_file by user, and set `init_score` field in dataset.

Reply:
another solution is to force the init_score_filename to `concat(data_filename, ".init"), and remove all initscore filenames parameters.